networkx/algorithms/centrality/tests/test_domirank.py

```diff
@@ -0,0 +1,197 @@
+        }
+        b, _, __ = nx.domirank(G, dt=0.5)
+        for n in sorted(G):
+            assert b[n] == pytest.approx(b_answer[n], abs=5e-3)
+
+    def test_P3_eps(self):
+        """DomiRank centrality: P3"""
+        G = nx.path_graph(3)
+        b_answer = {
+            0: -2.3477983474731445,
+            1: 4.496443748474121,
+            2: -2.3477983474731445,
+        }
+        b, _, bBool = nx.domirank(G, epsilon=1e-7)
+        assert bBool == True
+        for n in sorted(G):
+            assert b[n] == pytest.approx(b_answer[n], abs=5e-3)
+
+    def test_P3_max_iter(self):
+        """DomiRank centrality: P3"""
+        G = nx.path_graph(3)
+        b_answer = {0: -2.364804744720459, 1: 4.5204758644104, 2: -2.364804744720459}
+        b, _, bBool = nx.domirank(G, max_iter=5000)
+        for n in sorted(G):
+            assert b[n] == pytest.approx(b_answer[n], abs=5e-3)
+
+    def test_P3_check_step(self):
+        """DomiRank centrality: P3"""
+        G = nx.path_graph(3)
+        b_answer = {0: -2.346951484680176, 1: 4.495236396789551, 2: -2.346951484680176}
+        b, _, bBool = nx.domirank(G, check_step=1)
+        for n in sorted(G):
+            assert b[n] == pytest.approx(b_answer[n], abs=5e-3)
+
+
+class TestDomirankDirected:
+    def __init__(self):
+        self.G = nx.DiGraph()
+
+        edges = [
+            (1, 2),
+            (1, 3),
+            (2, 4),
+            (3, 2),
+            (3, 5),
+            (4, 2),
+            (4, 5),
+            (4, 6),
+            (5, 6),
+            (5, 7),
+            (5, 8),
+            (6, 8),
+            (7, 1),
+            (7, 5),
+            (7, 8),
+            (8, 6),
+            (8, 7),
+        ]
+
+        self.G.add_edges_from(edges)
+        self.G = self.G.reverse()
+        self.evc = {
+            1: 0.8088147044181824,
+            2: 1.3506824970245361,
+            3: 0.11302878707647324,
+            4: -0.20726482570171356,
+            5: 2.0464682579040527,
+            6: 0.3447999954223633,
+            7: -0.36868923902511597,
+            8: 0.5775589942932129,
+        }
+
+        self.evca = {
+            1: 0.8100015022345155,
+            2: 1.3502259832012713,
+            3: 0.11228691263288304,
+            4: -0.20697950162756873,
+            5: 2.0479405728029034,
+            6: 0.3427231477861121,
+            7: -0.3705877648939345,
+            8: 0.5791237659444093,
+        }
+
+    def test_domirank_centrality_directed(self):
+        G = self.G
+        b, _, bBool = nx.domirank(G)
+        assert bBool == True
+        for n in sorted(G):
+            assert b[n] == pytest.approx(self.evc[n], abs=5e-3)
+
+    def test_domirank_centrality_directed_analytical(self):
+        G = self.G
+        b, _, bBool = nx.domirank(G, analytical=True)
+        assert bBool == True
+        for n in sorted(G):
+            assert b[n] == pytest.approx(self.evca[n], abs=5e-3)
+
+
+class TestDomirankExceptions:
```

Comment:
You can probably clean this up with parametrized tests, using the [`@pytest.mark.parametrize` decorator](https://docs.pytest.org/en/stable/how-to/parametrize.html).
These [examples](https://docs.pytest.org/en/stable/example/parametrize.html#parametrizing-conditional-raising) even specifically cover testing conditional raising.

As a sidenote, [this section](https://docs.pytest.org/en/stable/example/parametrize.html#parametrization-of-optional-implementations-imports) above the one I linked to is also interesting re: the discussion on handling missing imports, which could mesh well with what I suggested for splitting up the no-numpy and numpy versions of the function and comparing their outputs.

Reply:
I have tried my best to use the @pytest.mark.parameterize decorator. It does indeed clean up the code quite a bit!

Reply:
Try something like this:
```python
class TestDomirankExceptions:
    @pytest.mark.parametrize(
        ("G", "analytical", "sigma", "expectation"),
        [
            (nx.Graph(), False, 0.95, pytest.raises(nx.NetworkXPointlessConcept)),
            (nx.path_graph(10), False, -1, pytest.raises(nx.NetworkXUnfeasible)),
            (nx.path_graph(10), False, 2, pytest.raises(nx.NetworkXUnfeasible)),
            (nx.path_graph(10), True, -1, pytest.raises(nx.NetworkXUnfeasible)),
            (nx.path_graph(10), False, 2, pytest.raises(nx.NetworkXUnfeasible)),
            (nx.MultiGraph(), False, 0.95, pytest.raises(nx.NetworkXNotImplemented)),
        ]
    )
    def test_domirank_exceptions(self, G, analytical, sigma, expectation):
        with expectation:
            nx.domirank(G, sigma=sigma, analytical=analytical)
```

Note that the exception tests are still missing some cases (e.g. all the various combinations of parameters that behave differently).

In all honesty, putting them all together in one function like I did here is probably a suboptimal way of doing these tests. It would be better to have separate parametrized test functions (parametrized in the relevant arguments for each) for the various semantically distinct tests:
- one for the pointless concept,
- one testing correct raising behavior for various graph types (using `from contextlib import nullcontext as does_not_raise()` as in the examples I linked to can help here,
- one for the interplay between `sigma` and `analytical`,
- etc.

Reply:
Wow! Thank you. I completely misunderstood their documentation. This makes a lot more sense!